src/leap/bitmask/services/soledad/soledadbootstrapper.py

```diff
@@ -666,36 +557,110 @@ def run_soledad_setup_checks(self, provider_config, user, password,
+            # XXX FIXME make this async too! (use txrequests)
+            # Also, why the fuck would we want to download it *every time*?
+            # We should be fine by using last-time config, or at least
+            # trying it.
+            self._download_config()
+            uuid = self.srpauth.get_uuid()
         except Exception as e:
             # TODO: we should handle more specific exceptions in here
             self._soledad = None
             self._keymanager = None
-            logger.exception("Error while bootstrapping Soledad: %r" % (e, ))
+            logger.exception("Error while bootstrapping Soledad: %r" % (e,))
             self._signaler.signal(signal_failed)
+            return defer.succeed(None)
+
+        # soledad config is ok, let's proceed to load and sync soledad
+        d = self.load_and_sync_soledad(uuid)
+        d.addCallback(lambda _: self._gen_key())
+        d.addCallback(lambda _: self._signaler.signal(signal_finished))
+        return d
+
+
+class Syncer:
+    """
+    Takes care of retries, timeouts and other issues while syncing
+    """
+    MAX_SYNC_RETRIES = 10
+    WAIT_MAX_SECONDS = 600
+
+    def __init__(self, soledad, signaler):
+        self._tries = 0
+        self._soledad = soledad
+        self._signaler = signaler
+
+    def sync(self):
+        self._callback_deferred = defer.Deferred()
+        self._try_sync()
+        return self._callback_deferred
+
+    def _try_sync(self):
+        logger.debug("BOOTSTRAPPER: trying to self Soledad....")
+        # pass defer_decryption=False to get inline decryption
+        # for debugging.
+        self._sync_deferred = self._soledad.sync(defer_decryption=True)
+        self._sync_deferred.addCallbacks(self._success, self._error)
+        reactor.callLater(self.WAIT_MAX_SECONDS, self._timeout)
+
+    def _success(self, result):
+        logger.debug("Soledad has been selfed!")
+        self._callback_deferred.callback(result)
+        # so long, and thanks for all the fish
+
+    def _error(self, failure):
+        if failure.check(InvalidAuthTokenError):
+            logger.error('Invalid auth token while trying to self Soledad')
+            self._signaler.signal(
+                self._signaler.soledad_invalid_auth_token)
+            self._callback_deferred.fail(failure)
+        elif failure.check(sqlite_ProgrammingError,
+                           sqlcipher_ProgrammingError):
+            logger.exception("%r" % (failure.value,))
+            self._callback_deferred.fail(failure)
+        elif failure.check(SSLError):
+            logger.error("%r" % (failure.value,))
+            self._retry()
+        elif failure.check(u1db_errors.InvalidGeneration):
+            logger.error("%r" % (failure.value,))
+            self._retry()
+        else:
+            logger.exception("Unhandled error while syncing "
+                             "soledad: %r" % (failure.value,))
+            self._retry()
+
+    def _timeout(self):
```

Comment:
this is nice, and way cleaner than the current implementation.
I wonder, for the future, thoght, if we shouldn't implement a timeout that can detect idle connection per sync-attempt (instead of this big monster timeout). We probably should also look into passing a cancellable to the soledad sync defer.
But again, not for now. We can live with this client-side hack for now.
Can you add a quick note on this class saying that all this logic should be moved to soledad itself (or at least accepting variable parameters for the timeouts/retries on init).